X.H.EwmhDesktops: optionally advertise fullscreen support in _NET_SUPPORTED

liskin opened the pull request:
Provide a way to advertise _NET_WM_STATE_FULLSCREEN in _NET_SUPPORTED.
I had this hardcoded for a while but read the mpv manpage recently:

```
Specifically, yes will force use of NetWM fullscreen support, even
if not advertised by the WM. This can be useful for WMs that are
broken on purpose, like XMonad. (XMonad supposedly doesn't advertise
fullscreen support, because Flash uses it. Apparently, applications
which want to use fullscreen anyway are supposed to either ignore
the NetWM support hints, or provide a workaround. Shame on XMonad
for deliberately breaking X protocols (as if X isn't bad enough
already).
```

I'm not sure Flash is still a good reason these days to not advertise
fullscreen support yet still handle it and look like idiots while doing so.
(And if anyone thinks it is a good reason, their unchanged xmonad config
will continue behaving that way.)

geekosaur commented:
On Sat, Oct 29, 2016 at 7:10 PM, Tomáš Janoušek <notifications@github.com>
wrote:

> I'm not sure Flash is still a good reason these days to not advertise
> fullscreen support yet still handle it and look like idiots while doing so.
> (And if anyone thinks it is a good reason, their unchanged xmonad config
> will continue behaving that way.)
>

Flash was never the reason, except indirectly (it's *part of* why we don't
wire support in unconditionally... but apparently offering mere users
configurability or optional features is Wrong these days; see also Gnome).
The real reason not to advertise is that the startupHook can't decompile
the handleEventHook to determine if a fullscreen event hook was present (C
can't do that either). But people will apparently always go for the
explanation that puts not-them in a bad light.

-- 
brandon s allbery kf8nh                               sine nomine associates
allbery.b@gmail.com                                  ballbery@sinenomine.net
unix, openafs, kerberos, infrastructure, xmonad        http://sinenomine.net

liskin commented:
On Sat, Oct 29, 2016 at 04:28:44PM -0700, geekosaur wrote:
> Flash was never the reason, except indirectly (it's *part of* why we don't
> wire support in unconditionally... but apparently offering mere users
> configurability or optional features is Wrong these days; see also Gnome).
> The real reason not to advertise is that the startupHook can't decompile
> the handleEventHook to determine if a fullscreen event hook was present (C
> can't do that either). But people will apparently always go for the
> explanation that puts not-them in a bad light.

If that was the reason, then possibly the X.H.EwmhDesktops API was less than
ideal. :-)

I mean, really, if xmonad in all those years couldn't figure out a way to make
handling and advertising fullscreen support configurable, then maybe it is the
one to blame, not mpv.

Anyway, that's the past. Here's a pull request that tries to fix this. I can
change the commit message, sure, but discussion about how GNOME approach is
bad and how developers blame everyone but themselves is out of scope here.
(Although I do understand your frustration.)

-- 
Tomáš Janoušek, a.k.a. Pivník, a.k.a. Liskni_si, http://work.lisk.in/

geekosaur commented:
On Sat, Oct 29, 2016 at 7:36 PM, Tomáš Janoušek <notifications@github.com>
wrote:

> If that was the reason, then possibly the X.H.EwmhDesktops API was less
> than
> ideal. :-)
>

The best way to do it is an extensible framework, so EwmhDesktops doesn't
have to be updated for new EWMH functionality (or old functionality not
currently supported/advertised, like fullscreen or minimize/restore).  I
wanted to try to do this without breaking existing uses, but as far as I
can tell that's not really viable.

That said, no, EwmhDesktops is fairly old and did not have any real thought
put into it with respect to flexibility or extensibility.

-- 
brandon s allbery kf8nh                               sine nomine associates
allbery.b@gmail.com                                  ballbery@sinenomine.net
unix, openafs, kerberos, infrastructure, xmonad        http://sinenomine.net

liskin commented:
On Sat, Oct 29, 2016 at 05:21:13PM -0700, geekosaur wrote:
> The best way to do it is an extensible framework, so EwmhDesktops doesn't
> have to be updated for new EWMH functionality (or old functionality not
> currently supported/advertised, like fullscreen or minimize/restore).  I
> wanted to try to do this without breaking existing uses, but as far as I
> can tell that's not really viable.

One option that comes to mind is xmonad (or, well, whatever, really) clearing
_NET_WM_SUPPORTED on startup and contrib modules adding things and handling
them. Right now EwmhDesktops.ewmh resets _NET_WM_SUPPORTED and wm name so its
hooks must run before everything else.

It's not really an "extensible framework", just a hack to make things work but
I think it could work without breaking existing uses. Am I overlooking
something?

-- 
Tomáš Janoušek, a.k.a. Pivník, a.k.a. Liskni_si, http://work.lisk.in/

geekosaur commented:
On Sat, Oct 29, 2016 at 8:26 PM, Tomáš Janoušek <notifications@github.com>
wrote:

> It's not really an "extensible framework", just a hack to make things work
> but
> I think it could work without breaking existing uses. Am I overlooking
> something?
>

My current thought for a quick hackaround (since I'm unlikely to be doing
the real thing any time soon) is `ewmhWith [atom,...]` and each module that
offers an extension (including `EwmhDesktops` itself for fullscreen)
exporting a binding for the relevant `_NET_SUPPORTED` atom. So if you're
using the fullscreen and minimize/restore hooks, you'd use ewmhWith
[eWithFullScreen, eWithMinimize] and those two would be defined as
respectively `getAtom "_NET_WM_STATE_FULLSCREEN"` and `getAtom
"_NET_WM_STATE_HIDDEN"`. This would be an alternative to the `ewmh`
combinator applied to a config, or available as `ewmhStartupHookWith` for
old-style. (We're actively tossing this around in `#xmonad` right now.)

-- 
brandon s allbery kf8nh                               sine nomine associates
allbery.b@gmail.com                                  ballbery@sinenomine.net
unix, openafs, kerberos, infrastructure, xmonad        http://sinenomine.net

geekosaur commented:
Unfortunately, the fact that it took 11 days for me to respond to this is a
good measure of how likely it is that I'd get anything going for a proper
solution. :/

On Wed, Nov 9, 2016 at 6:27 PM, Brandon Allbery <allbery.b@gmail.com> wrote:

>
> On Sat, Oct 29, 2016 at 8:26 PM, Tomáš Janoušek <notifications@github.com>
> wrote:
>
>> It's not really an "extensible framework", just a hack to make things
>> work but
>> I think it could work without breaking existing uses. Am I overlooking
>> something?
>>
>
> My current thought for a quick hackaround (since I'm unlikely to be doing
> the real thing any time soon) is `ewmhWith [atom,...]` and each module that
> offers an extension (including `EwmhDesktops` itself for fullscreen)
> exporting a binding for the relevant `_NET_SUPPORTED` atom. So if you're
> using the fullscreen and minimize/restore hooks, you'd use ewmhWith
> [eWithFullScreen, eWithMinimize] and those two would be defined as
> respectively `getAtom "_NET_WM_STATE_FULLSCREEN"` and `getAtom
> "_NET_WM_STATE_HIDDEN"`. This would be an alternative to the `ewmh`
> combinator applied to a config, or available as `ewmhStartupHookWith` for
> old-style. (We're actively tossing this around in `#xmonad` right now.)
>
> --
> brandon s allbery kf8nh                               sine nomine
> associates
> allbery.b@gmail.com
> ballbery@sinenomine.net
> unix, openafs, kerberos, infrastructure, xmonad
> http://sinenomine.net
>



-- 
brandon s allbery kf8nh                               sine nomine associates
allbery.b@gmail.com                                  ballbery@sinenomine.net
unix, openafs, kerberos, infrastructure, xmonad        http://sinenomine.net

liskin commented:
Yeah, that seems like a good idea. Before I submitted this pull request I thought about a config parameter to `ewmh` but then I noticed `XMonad.Hooks.SetWMName` and decided against it. But rewriting everything EWMH-related to an extensible framework like `ewmhWith` might be a way to go. I'll try to code something up.

geekosaur commented:
That's not the true extensible one, though. I wanted one where you could
also address, for example, the window focus one with e.g. `ewmhHandler
"_NET_ACTIVE_WINDOW" (const ())` which would advertise focus switching (so
the program doesn't fallback to just stealing focus itself) but ignore it.
(This is not actually what it would need to look like, because some things
would be handleEventHooks and others are logHook, etc.) Or the
EWMH-sanctioned version of that, which instead of completely ignoring it
converts it into an urgency marker (see X.H.UrgencyHooks). This minimal
version just lets us control advertisements.

The big problem with the proper extensible framework is that it pretty much
has to be a combinator and you an't modify the hooks directly any more for
EWMH. I had wanted to be able to add stuff to startupHook, but that
requires a persistent state that can cope with functions whereas ours
necessarily requires a Show instance to persist across xmonad instances.

On Thu, Nov 10, 2016 at 3:45 AM, Tomáš Janoušek <notifications@github.com>
wrote:

> Yeah, that seems like a good idea. Before I submitted this pull request I
> thought about a config parameter to ewmh but then I noticed
> XMonad.Hooks.SetWMName and decided against it. But rewriting everything
> EWMH-related to an extensible framework like ewmhWith might be a way to
> go. I'll try to code something up.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/xmonad/xmonad-contrib/pull/109#issuecomment-259633095>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB8SoGtgkf--7APLy2XXPYgDqpCDafZCks5q8tmtgaJpZM4KkPid>
> .
>



-- 
brandon s allbery kf8nh                               sine nomine associates
allbery.b@gmail.com                                  ballbery@sinenomine.net
unix, openafs, kerberos, infrastructure, xmonad        http://sinenomine.net

geekosaur commented:
Oh, and to be clear, extension modules would expose "canned" applications
of `ewmhHandler` rather than make the user figure it out, so the user would
`ewmhIgnoreFocusChange` instead of `ewmhHandler ...`). ewmhHandler would be
exposed for modules to use in implementing this, or for advanced users who
want to e.g. experiment with newer EWMH additions (the spec is modified
fairly regularly).

On Thu, Nov 10, 2016 at 1:21 PM, Brandon Allbery <allbery.b@gmail.com>
wrote:

> That's not the true extensible one, though. I wanted one where you could
> also address, for example, the window focus one with e.g. `ewmhHandler
> "_NET_ACTIVE_WINDOW" (const ())` which would advertise focus switching (so
> the program doesn't fallback to just stealing focus itself) but ignore it.
> (This is not actually what it would need to look like, because some things
> would be handleEventHooks and others are logHook, etc.) Or the
> EWMH-sanctioned version of that, which instead of completely ignoring it
> converts it into an urgency marker (see X.H.UrgencyHooks). This minimal
> version just lets us control advertisements.
>
> The big problem with the proper extensible framework is that it pretty
> much has to be a combinator and you an't modify the hooks directly any more
> for EWMH. I had wanted to be able to add stuff to startupHook, but that
> requires a persistent state that can cope with functions whereas ours
> necessarily requires a Show instance to persist across xmonad instances.
>
> On Thu, Nov 10, 2016 at 3:45 AM, Tomáš Janoušek <notifications@github.com>
> wrote:
>
>> Yeah, that seems like a good idea. Before I submitted this pull request I
>> thought about a config parameter to ewmh but then I noticed
>> XMonad.Hooks.SetWMName and decided against it. But rewriting everything
>> EWMH-related to an extensible framework like ewmhWith might be a way to
>> go. I'll try to code something up.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/xmonad/xmonad-contrib/pull/109#issuecomment-259633095>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AB8SoGtgkf--7APLy2XXPYgDqpCDafZCks5q8tmtgaJpZM4KkPid>
>> .
>>
>
>
>
> --
> brandon s allbery kf8nh                               sine nomine
> associates
> allbery.b@gmail.com
> ballbery@sinenomine.net
> unix, openafs, kerberos, infrastructure, xmonad
> http://sinenomine.net
>



-- 
brandon s allbery kf8nh                               sine nomine associates
allbery.b@gmail.com                                  ballbery@sinenomine.net
unix, openafs, kerberos, infrastructure, xmonad        http://sinenomine.net

pjones commented:
Where do we stand on this PR @geekosaur?

geekosaur commented:
I'm not going to be writing anything in the near future; I'm still trying to get scheduled for sinus/nasal surgery (and hoping it doesn't interact with my "wonderful" brain the same way the sinus/nasal issues are...).

sgf-dma commented:
Following the recent discussion of [GLFW fullscreen](https://mail.haskell.org/pipermail/xmonad/2017-March/015224.html) on xmonad maillist, can you merge this? Also, we need to advertise `_NET_WM_STATE` in `_NET_SUPPORTED` (see discussion linked above), but that may be another PR.

This solution may not be ideal from the point of better EWMH support rewrite, but for a time being (and for a long time past) `fullscreenEventHook` is the only solution available

pjones commented:
Please rebase this and add an entry in `CHANGES.md`.

cstrahan commented:
FWIW, I also came here from the same mailing list thread that @sgf-dma pointed out. In my case, [`sxiv`](https://github.com/muennich/sxiv) was refusing to enter fullscreen with "sxiv: No fullscreen support".

sgf-dma commented:
@cstrahan , does the patch [mentioned in mailing](https://gist.github.com/sgf-dma/a609f855bbacf1a0292e660c32a5a04e) list help?

cstrahan commented:
@sgf-dma Yeah. Here's what I'm using:

https://github.com/cstrahan/dotfiles/blob/b28910dca345f5030969828fae1fdbb9982e0042/.xmonad/xmonad.hs#L64-L82

troydm commented:
I hope this will be merged someday 👍

geekosaur commented:
Does anyone plan to address the git conflicts in this?

liskin commented:
Yes they do! :-)

Note that I just rebased it, I didn't run this code for any period of time. I don't expect it to break, the conflicts were trivial, though.

sgf-dma commented:
@liskin  , Can you also advertise `_NET_WM_STATE` as noted in [comment above](https://github.com/xmonad/xmonad-contrib/pull/109#issuecomment-284985832), because you patch is still not enough for some programs to work properly.

liskin commented:
@pjones @sgf-dma I updated the PR, it now advertises `_NET_WM_STATE` as well and I added an entry to `CHANGES.md`. I think it's finally ready to be merged.

(Sorry it took so long. Please do feel free to ping me on #xmonad@freenode next time I forget about a PR. I'd have been totally okay with someone else pushing into my branches should it speed things up. :-))

liskin commented:
@byorgey @peti This one's ready again (i.e. both 9.needs labels may be dropped) and includes a fix for the misplaced CHANGES entries I commented about a few minutes ago.

geekosaur commented:
I suspect the unrelated `CHANGES.md` changes should be their own PR, instead of being part of this one.

liskin commented:
Yeah, whatever, feel free to `git cherry-pick --reset-author` even.

liskin commented:
But then, I can get Hacktoberfest points for this. Let's do it. :-D

geekosaur commented:
I'm not going to merge this one because I haven't been paying close enough attention, but the other was easy enough to verify and getting it merged should simplify merging this one.

byorgey commented:
To be honest I don't really understand the issues here, but seeing as several people have commented over the years asking for this to be merged, no one has voiced objections, and it passes CI, I'm going to merge it.  I'm also adding @liskin to the xmonad-contrib team.
